spec/controllers/reports/regions_controller_spec.rb

```diff
@@ -550,37 +573,163 @@
+    let(:region) { facility.region.district_region }
+
+    it "returns 401 when user is not authorized" do
+      facility
+
+      get :diabetes_monthly_district_data, params: {id: region.slug, report_scope: "district", format: "csv"}
+      expect(response.status).to eq(401)
+    end
+
+    it "returns 302 found with invalid region" do
+      facility
+      sign_in(cvho.email_authentication)
+
+      get :diabetes_monthly_district_data, params: {id: "not-found", report_scope: "district", format: "csv"}
+      expect(response.status).to eq(302)
+    end
+
+    it "returns 302 if region is not district" do
+      facility
+      sign_in(cvho.email_authentication)
+
+      get :diabetes_monthly_district_data, params: {id: facility.slug, report_scope: "district", format: "csv"}
+      expect(response.status).to eq(302)
+    end
+
+    it "calls csv service and returns 200 with csv data" do
+      Timecop.freeze("June 15th 2020") do
+        facility
+        sign_in(cvho.email_authentication)
+
+        expect_any_instance_of(MonthlyDistrictData::Exporter).to receive(:report).and_call_original
+        get :diabetes_monthly_district_data, params: {id: region.slug, report_scope: "district", format: "csv"}
+        expect(response.status).to eq(200)
+        expect(response.body).to include("Monthly facility data for #{region.name} #{Date.current.strftime("%B %Y")}")
+        report_date = Date.current.strftime("%b-%Y").downcase
+        expected_filename = "monthly-facility-data-#{region.slug}-#{report_date}.csv"
+        expect(response.headers["Content-Disposition"]).to include(%(filename="#{expected_filename}"))
+      end
+    end
+
+    it "passes the provided period to the csv service" do
+      facility
+      sign_in(cvho.email_authentication)
+
+      period = Period.month("July 2018")
+      exporter = MonthlyDistrictData::Diabetes.new(
+        region: region,
+        period: period,
+        medications_dispensation_enabled: false
+      )
+      allow(MonthlyDistrictData::Diabetes).to receive(:new).and_return(exporter)
+      expect(MonthlyDistrictData::Exporter).to receive(:new).with(exporter: exporter).and_call_original
+      get :diabetes_monthly_district_data,
+        params: {id: region.slug, report_scope: "district", format: "csv", period: period.attributes}
+    end
+  end
+
+  describe "#hypertension_monthly_state_data" do
+    let(:facility_group) { create(:facility_group, organization: organization) }
+    let(:facility) { create(:facility, facility_group: facility_group) }
+    let(:region) { facility.region.state_region }
+
+    it "returns 401 when user is not authorized" do
+      facility
+
+      get :hypertension_monthly_state_data, params: {id: region.slug, report_scope: "state", format: "csv"}
+      expect(response.status).to eq(401)
+    end
+
+    it "returns 302 found with invalid region" do
+      facility
+      sign_in(cvho.email_authentication)
```

Comment:
[Style] You could wrap all the authenticated blocks in a single context, and `sign_in` in a `before` block.

Reply:
done